### 모듈 불러오기

In [19]:
import pandas as pd
from pandas import Series, DataFrame
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
import seaborn as sns

##### 데이터 불러오기

In [20]:
import os
os.listdir("C:/Users/gun67/전력 프로젝트/2013")

['a. 2013_1_House1_Ch1_Combined_Data_New.csv',
 'b. 2013_2_House1_Ch1_Combined_Data_New.csv',
 'c. 2013_3_House1_Ch1_Combined_Data_New.csv',
 'd. 2013_4_House1_Ch1_Combined_Data_New.csv',
 'e. 2013_5_House1_Ch1_Combined_Data_New.csv',
 'f. 2013_6_House1_Ch1_Combined_Data_New.csv',
 'g. 2013_7_House1_Ch1_Combined_Data_New.csv',
 'h. 2013_8_House1_Ch1_Combined_Data_New.csv',
 'i. 2013_9_House1_Ch1_Combined_Data_New.csv',
 'j. 2013_10_House1_Ch1_Combined_Data_New.csv',
 'k. 2013_11_House1_Ch1_Combined_Data_New.csv',
 'l. 2013_12_House1_Ch1_Combined_Data_New.csv']

In [21]:
cd C:/Users/gun67/전력 프로젝트/2013

C:\Users\gun67\전력 프로젝트\2013


In [22]:
forders = os.listdir("C:/Users/gun67/전력 프로젝트/2013")
df_all_data = pd.DataFrame()

for files in forders :
    df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
    df_all_data = pd.concat([df_all_data, df], axis = 0)

C:\Users\gun67\AppData\Local\Temp\ipykernel_4784\1368539699.py:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
C:\Users\gun67\AppData\Local\Temp\ipykernel_4784\1368539699.py:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
C:\Users\gun67\AppData\Local\Temp\ipykernel_4784\1368539699.py:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df = pd.read_csv(files, dtype = {'Code':str}, parse_dates = ["Watt"])
C:\Users\gun67\AppData\Local\Temp\

In [23]:
df_all_data

Unnamed: 0 Year-Month-Date       Time  Holiday Watt  Temp('C)  \
0               0      2013-01-01  [ 00:00 ]        1  522      10.0   
1               1      2013-01-01  [ 00:01 ]        1  558      10.0   
2               2      2013-01-01  [ 00:02 ]        1  537      10.0   
3               3      2013-01-01  [ 00:03 ]        1  525      10.0   
4               4      2013-01-01  [ 00:04 ]        1  539      10.0   
...           ...             ...        ...      ...  ...       ...   
44635       44635      2013-12-31  [ 23:55 ]        0  401       8.0   
44636       44636      2013-12-31  [ 23:56 ]        0  400       8.0   
44637       44637      2013-12-31  [ 23:57 ]        0  400       8.0   
44638       44638      2013-12-31  [ 23:58 ]        0  497       8.0   
44639       44639      2013-12-31  [ 23:59 ]        0  493       8.0   

       Humidity(%)  CPI  
0           87.000  2.7  
1           86.833  2.7  
2           86.667  2.7  
3           86.500  2.7  
4           86.333  2.7  
...            ...  ...  
44635       81.000  2.0  
44636       81.000  2.0  
44637       81.000  2.0  
44638       81.000  2.0  
44639       81.000  2.0  

[525600 rows x 8 columns]

### 불필요한 열 제거

In [24]:
df_all_data.drop(columns = ["Unnamed: 0", "CPI"], inplace = True)
df_all_data

Year-Month-Date       Time  Holiday Watt  Temp('C)  Humidity(%)
0          2013-01-01  [ 00:00 ]        1  522      10.0       87.000
1          2013-01-01  [ 00:01 ]        1  558      10.0       86.833
2          2013-01-01  [ 00:02 ]        1  537      10.0       86.667
3          2013-01-01  [ 00:03 ]        1  525      10.0       86.500
4          2013-01-01  [ 00:04 ]        1  539      10.0       86.333
...               ...        ...      ...  ...       ...          ...
44635      2013-12-31  [ 23:55 ]        0  401       8.0       81.000
44636      2013-12-31  [ 23:56 ]        0  400       8.0       81.000
44637      2013-12-31  [ 23:57 ]        0  400       8.0       81.000
44638      2013-12-31  [ 23:58 ]        0  497       8.0       81.000
44639      2013-12-31  [ 23:59 ]        0  493       8.0       81.000

[525600 rows x 6 columns]

### 인덱스 초기화

In [25]:
df_all_data.reset_index(drop = True, inplace = True)
df_all_data

Year-Month-Date       Time  Holiday Watt  Temp('C)  Humidity(%)
0           2013-01-01  [ 00:00 ]        1  522      10.0       87.000
1           2013-01-01  [ 00:01 ]        1  558      10.0       86.833
2           2013-01-01  [ 00:02 ]        1  537      10.0       86.667
3           2013-01-01  [ 00:03 ]        1  525      10.0       86.500
4           2013-01-01  [ 00:04 ]        1  539      10.0       86.333
...                ...        ...      ...  ...       ...          ...
525595      2013-12-31  [ 23:55 ]        0  401       8.0       81.000
525596      2013-12-31  [ 23:56 ]        0  400       8.0       81.000
525597      2013-12-31  [ 23:57 ]        0  400       8.0       81.000
525598      2013-12-31  [ 23:58 ]        0  497       8.0       81.000
525599      2013-12-31  [ 23:59 ]        0  493       8.0       81.000

[525600 rows x 6 columns]

### 데이터 타입 확인

In [26]:
print(df_all_data.dtypes)

Year-Month-Date     object
Time                object
Holiday              int64
Watt                object
Temp('C)           float64
Humidity(%)        float64
dtype: object


##### 숫자형 데이터로 표시되야 할 Watt 열에 잘못 기입된 데이터가 있다고 판단  
##### 데이터 타입을 숫자형으로 바꾸기 테스트 진행

In [27]:
df_all_data["Watt"] = pd.to_numeric(df_all_data["Watt"], errors = 'coerce')
print(df_all_data.dtypes)

Year-Month-Date     object
Time                object
Holiday              int64
Watt                 int64
Temp('C)           float64
Humidity(%)        float64
dtype: object


##### 다행히 Watt 열의 데이터 타입이 수치형으로 변경되었음을 알 수 있음

### 결손치 확인

In [28]:
print(df_all_data[df_all_data['Watt'] == 0])
print("-----------------------------------------")
print(df_all_data[df_all_data["Temp('C)"] == "NaN"])
print("-----------------------------------------")
print(df_all_data[(df_all_data["Humidity(%)"] < 0) | (df_all_data["Humidity(%)"] > 100)])

       Year-Month-Date       Time  Holiday  Watt  Temp('C)  Humidity(%)
1044        2013-01-01  [ 17:24 ]        1     0     6.000       71.200
4744        2013-01-04  [ 07:04 ]        0     0     9.867       76.800
5080        2013-01-04  [ 12:40 ]        0     0    10.000       76.000
6644        2013-01-05  [ 14:44 ]        1     0    10.000       84.667
6645        2013-01-05  [ 14:45 ]        1     0    10.000       84.500
...                ...        ...      ...   ...       ...          ...
196354      2013-05-17  [ 08:34 ]        0     0     9.133       80.533
221463      2013-06-03  [ 19:03 ]        0     0    13.000       63.000
221464      2013-06-03  [ 19:04 ]        0     0    13.000       63.000
221465      2013-06-03  [ 19:05 ]        0     0    13.000       63.000
221466      2013-06-03  [ 19:06 ]        0     0    13.000       63.000

[34449 rows x 6 columns]
-----------------------------------------
Empty DataFrame
Columns: [Year-Month-Date, Time, Holiday, Watt, Temp

##### Watt 열에서는 525600개의 레코드 중 34449개, 약 5.6%의 결손치가 확인 됨  
##### Temp('C) 열과 Humidicy(%) 열은 결손치가 없다고 판단

### 결손치 처리

In [29]:
# 전력값 결손치를 평균값으로 대체
data_watt = pd.DataFrame(df_all_data["Watt"])
print("결손치 대체 전 데이터")
print(data_watt)

# 0이 아닌 값을 가지는 데이터의 평균 계산
Watt_mean = data_watt[data_watt["Watt"] != 0]["Watt"].mean()

# 결손치를 평균값으로 대체
df_all_data.loc[data_watt["Watt"] == 0, "Watt"] = Watt_mean

print("\n결손치 대체 후 데이터")
print(df_all_data)

결손치 대체 전 데이터
        Watt
0        522
1        558
2        537
3        525
4        539
...      ...
525595   401
525596   400
525597   400
525598   497
525599   493

[525600 rows x 1 columns]

결손치 대체 후 데이터
       Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
0           2013-01-01  [ 00:00 ]        1  522.0      10.0       87.000
1           2013-01-01  [ 00:01 ]        1  558.0      10.0       86.833
2           2013-01-01  [ 00:02 ]        1  537.0      10.0       86.667
3           2013-01-01  [ 00:03 ]        1  525.0      10.0       86.500
4           2013-01-01  [ 00:04 ]        1  539.0      10.0       86.333
...                ...        ...      ...    ...       ...          ...
525595      2013-12-31  [ 23:55 ]        0  401.0       8.0       81.000
525596      2013-12-31  [ 23:56 ]        0  400.0       8.0       81.000
525597      2013-12-31  [ 23:57 ]        0  400.0       8.0       81.000
525598      2013-12-31  [ 23:58 ]        0  497.0       8.0 

C:\Users\gun67\AppData\Local\Temp\ipykernel_4784\280170690.py:10: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '411.36976204873855' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  df_all_data.loc[data_watt["Watt"] == 0, "Watt"] = Watt_mean


### 이상치 확인

In [30]:
# Watt

# 통계적 방법으로 이상치 확인
# Z-점수 ; 각 데이터 포인트가 평균에서 몇 표준편차 떨어져 있는지 나타냄, 일반적으로 Z-점수가 3 이상이거나 -3 이하인 경우 이상치로 간주
# 사분위 범위 ; Q1(1사분위수)와 Q3(3사분위수) 사이의 범위를 계산하고, Q1 - 1.5 * IQR 보다 작거나 Q3 + 1.5 * IQR 보다 큰 값을 이상치로 간주

# Z-점수 계산
z_scores = np.abs(stats.zscore(df_all_data["Watt"]))
print("Z-scores : ", z_scores)

# IQR 계산
Q1 = df_all_data["Watt"].quantile(0.25)
Q3 = df_all_data["Watt"].quantile(0.75)
IQR = Q3 - Q1
print("IQR : ", IQR)

# 이상치 확인
outliers = df_all_data[(df_all_data["Watt"] < (Q1 - 1.5 * IQR)) | (df_all_data["Watt"] > (Q3 + 1.5 * IQR))]
print("Outliers based on IQR : ", outliers)

# z_scores 가 3 미만인 데이터 찾기
cleaned_data_watt = df_all_data[(z_scores < 3)]
print("Data after removing outliers based on Z-score : ", cleaned_data_watt)

Z-scores :  0         0.218254
1         0.289276
2         0.247847
3         0.224173
4         0.251793
            ...   
525595    0.020458
525596    0.022431
525597    0.022431
525598    0.168934
525599    0.161042
Name: Watt, Length: 525600, dtype: float64
IQR :  218.36976204873855
Outliers based on IQR :         Year-Month-Date       Time  Holiday    Watt  Temp('C)  Humidity(%)
27          2013-01-01  [ 00:27 ]        1  1136.0    10.000       82.500
28          2013-01-01  [ 00:28 ]        1  1136.0    10.000       82.333
55          2013-01-01  [ 00:55 ]        1  2483.0     9.167       86.167
56          2013-01-01  [ 00:56 ]        1  2492.0     9.133       86.333
57          2013-01-01  [ 00:57 ]        1  2488.0     9.100       86.500
...                ...        ...      ...     ...       ...          ...
525368      2013-12-31  [ 20:08 ]        0   745.0     7.000       76.000
525372      2013-12-31  [ 20:12 ]        0   741.0     7.000       76.000
525373      2013-12

In [31]:
# Temp('C)

# Z-점수 계산
z_scores = np.abs(stats.zscore(df_all_data["Temp('C)"]))
print("Z-scores : ", z_scores)

# IQR 계산
Q1 = df_all_data["Temp('C)"].quantile(0.25)
Q3 = df_all_data["Temp('C)"].quantile(0.75)
IQR = Q3 - Q1
print("IQR : ", IQR)

# 이상치 확인
outliers = df_all_data[(df_all_data["Temp('C)"] < (Q1 - 1.5 * IQR)) | (df_all_data["Temp('C)"] > (Q3 + 1.5 * IQR))]
print("Outliers based on IQR : ", outliers)

# z_scores 가 3 미만인 데이터 찾기
cleaned_df_Temp = df_all_data[(z_scores < 3)]
print("Data after removing outliers based on Z-score : ", cleaned_df_Temp)

Z-scores :  0         0.177163
1         0.177163
2         0.177163
3         0.177163
4         0.177163
            ...   
525595    0.481655
525596    0.481655
525597    0.481655
525598    0.481655
525599    0.481655
Name: Temp('C), Length: 525600, dtype: float64
IQR :  10.0
Outliers based on IQR :         Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
291736      2013-07-22  [ 14:16 ]        0  196.0    31.067         38.6
291737      2013-07-22  [ 14:17 ]        0  199.0    31.133         38.2
291738      2013-07-22  [ 14:18 ]        0  199.0    31.200         37.8
291739      2013-07-22  [ 14:19 ]        0  143.0    31.267         37.4
291740      2013-07-22  [ 14:20 ]        0  143.0    31.333         37.0
...                ...        ...      ...    ...       ...          ...
306354      2013-08-01  [ 17:54 ]        0  288.0    31.200         42.4
306355      2013-08-01  [ 17:55 ]        0  288.0    31.167         42.5
306356      2013-08-01  [ 17:56 ]     

In [32]:
# humidity(%)

# Z-점수 계산
z_scores = np.abs(stats.zscore(df_all_data["Humidity(%)"]))
print("Z-scores : ", z_scores)

# IQR 계산
Q1 = df_all_data["Humidity(%)"].quantile(0.25)
Q3 = df_all_data["Humidity(%)"].quantile(0.75)
IQR = Q3 - Q1
print("IQR : ", IQR)

# 이상치 확인
outliers = df_all_data[(df_all_data["Humidity(%)"] < (Q1 - 1.5 * IQR)) | (df_all_data["Humidity(%)"] > (Q3 + 1.5 * IQR))]
print("Outliers based on IQR : ", outliers)

# z_scores 가 3 미만인 데이터 찾기
cleaned_df_humidity = df_all_data[(z_scores < 3)]
print("Data after removing outliers based on Z-score : ", cleaned_df_humidity)

Z-scores :  0         0.873060
1         0.861652
2         0.850313
3         0.838906
4         0.827498
            ...   
525595    0.463212
525596    0.463212
525597    0.463212
525598    0.463212
525599    0.463212
Name: Humidity(%), Length: 525600, dtype: float64
IQR :  21.599999999999994
Outliers based on IQR :         Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
138986      2013-04-07  [ 12:26 ]        1  266.0     9.000       32.667
138987      2013-04-07  [ 12:27 ]        1  252.0     9.000       32.500
138988      2013-04-07  [ 12:28 ]        1  251.0     9.000       32.333
138989      2013-04-07  [ 12:29 ]        1  237.0     9.000       32.000
138990      2013-04-07  [ 12:30 ]        1  235.0     9.000       32.000
...                ...        ...      ...    ...       ...          ...
291969      2013-07-22  [ 18:09 ]        0  262.0    30.700       32.200
291970      2013-07-22  [ 18:10 ]        0  264.0    30.667       32.333
291971      2013-07-2

### 이상치 처리

In [33]:
# Watt

# 이상치 대체 ; 평균, 중앙값 등으로 대체
# 이상치를 중앙값으로 대체
median_value = df_all_data["Watt"].median()
df_all_data.loc[(z_scores >= 3), "Watt"] = median_value
print("Data after replacing outliers with median value : ", df_all_data)

Data after replacing outliers with median value :         Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
0           2013-01-01  [ 00:00 ]        1  522.0      10.0       87.000
1           2013-01-01  [ 00:01 ]        1  558.0      10.0       86.833
2           2013-01-01  [ 00:02 ]        1  537.0      10.0       86.667
3           2013-01-01  [ 00:03 ]        1  525.0      10.0       86.500
4           2013-01-01  [ 00:04 ]        1  539.0      10.0       86.333
...                ...        ...      ...    ...       ...          ...
525595      2013-12-31  [ 23:55 ]        0  401.0       8.0       81.000
525596      2013-12-31  [ 23:56 ]        0  400.0       8.0       81.000
525597      2013-12-31  [ 23:57 ]        0  400.0       8.0       81.000
525598      2013-12-31  [ 23:58 ]        0  497.0       8.0       81.000
525599      2013-12-31  [ 23:59 ]        0  493.0       8.0       81.000

[525600 rows x 6 columns]


In [34]:
# Temp

# 이상치를 중앙값으로 대체
median_value = df_all_data["Temp('C)"].median()
df_all_data.loc[(z_scores >= 3), "Temp('C)"] = median_value
print("Data after replacing outliers with median value : ", df_all_data)

Data after replacing outliers with median value :         Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
0           2013-01-01  [ 00:00 ]        1  522.0      10.0       87.000
1           2013-01-01  [ 00:01 ]        1  558.0      10.0       86.833
2           2013-01-01  [ 00:02 ]        1  537.0      10.0       86.667
3           2013-01-01  [ 00:03 ]        1  525.0      10.0       86.500
4           2013-01-01  [ 00:04 ]        1  539.0      10.0       86.333
...                ...        ...      ...    ...       ...          ...
525595      2013-12-31  [ 23:55 ]        0  401.0       8.0       81.000
525596      2013-12-31  [ 23:56 ]        0  400.0       8.0       81.000
525597      2013-12-31  [ 23:57 ]        0  400.0       8.0       81.000
525598      2013-12-31  [ 23:58 ]        0  497.0       8.0       81.000
525599      2013-12-31  [ 23:59 ]        0  493.0       8.0       81.000

[525600 rows x 6 columns]


In [35]:
# Humidity

# 이상치를 중앙값으로 대체
median_value = df_all_data["Humidity(%)"].median()
df_all_data.loc[(z_scores >= 3), "Humidity(%)"] = median_value
print("Data after replacing outliers with median value : ", df_all_data)

Data after replacing outliers with median value :         Year-Month-Date       Time  Holiday   Watt  Temp('C)  Humidity(%)
0           2013-01-01  [ 00:00 ]        1  522.0      10.0       87.000
1           2013-01-01  [ 00:01 ]        1  558.0      10.0       86.833
2           2013-01-01  [ 00:02 ]        1  537.0      10.0       86.667
3           2013-01-01  [ 00:03 ]        1  525.0      10.0       86.500
4           2013-01-01  [ 00:04 ]        1  539.0      10.0       86.333
...                ...        ...      ...    ...       ...          ...
525595      2013-12-31  [ 23:55 ]        0  401.0       8.0       81.000
525596      2013-12-31  [ 23:56 ]        0  400.0       8.0       81.000
525597      2013-12-31  [ 23:57 ]        0  400.0       8.0       81.000
525598      2013-12-31  [ 23:58 ]        0  497.0       8.0       81.000
525599      2013-12-31  [ 23:59 ]        0  493.0       8.0       81.000

[525600 rows x 6 columns]


### 데이터셋 통합

In [36]:
Watt_N = pd.DataFrame(df_all_data["Watt"], columns = ["Watt"])
Temp_N = pd.DataFrame(df_all_data["Temp('C)"], columns = ["Temp('C)"])
Humidity_N = pd.DataFrame(df_all_data["Humidity(%)"], columns = ["Humidity(%)"])

In [37]:
df_all = pd.concat([df_all_data["Year-Month-Date"], df_all_data["Time"], Watt_N, Temp_N, Humidity_N], axis = 1)
df_all

Year-Month-Date       Time   Watt  Temp('C)  Humidity(%)
0           2013-01-01  [ 00:00 ]  522.0      10.0       87.000
1           2013-01-01  [ 00:01 ]  558.0      10.0       86.833
2           2013-01-01  [ 00:02 ]  537.0      10.0       86.667
3           2013-01-01  [ 00:03 ]  525.0      10.0       86.500
4           2013-01-01  [ 00:04 ]  539.0      10.0       86.333
...                ...        ...    ...       ...          ...
525595      2013-12-31  [ 23:55 ]  401.0       8.0       81.000
525596      2013-12-31  [ 23:56 ]  400.0       8.0       81.000
525597      2013-12-31  [ 23:57 ]  400.0       8.0       81.000
525598      2013-12-31  [ 23:58 ]  497.0       8.0       81.000
525599      2013-12-31  [ 23:59 ]  493.0       8.0       81.000

[525600 rows x 5 columns]

### 데이터 저장

In [38]:
df_all.to_csv("C:/Users/gun67/전력 프로젝트/preprocessing_data/2013_preprocessing_data", index = False, encoding = "utf-8-sig")

### 저장 데이터 확인

In [39]:
pd.read_csv("C:/Users/gun67/전력 프로젝트/preprocessing_data/2013_preprocessing_data", encoding = "utf-8-sig")

Year-Month-Date       Time   Watt  Temp('C)  Humidity(%)
0           2013-01-01  [ 00:00 ]  522.0      10.0       87.000
1           2013-01-01  [ 00:01 ]  558.0      10.0       86.833
2           2013-01-01  [ 00:02 ]  537.0      10.0       86.667
3           2013-01-01  [ 00:03 ]  525.0      10.0       86.500
4           2013-01-01  [ 00:04 ]  539.0      10.0       86.333
...                ...        ...    ...       ...          ...
525595      2013-12-31  [ 23:55 ]  401.0       8.0       81.000
525596      2013-12-31  [ 23:56 ]  400.0       8.0       81.000
525597      2013-12-31  [ 23:57 ]  400.0       8.0       81.000
525598      2013-12-31  [ 23:58 ]  497.0       8.0       81.000
525599      2013-12-31  [ 23:59 ]  493.0       8.0       81.000

[525600 rows x 5 columns]